app/api/contests/enter.go

```diff
@@ -0,0 +1,110 @@
+package contests
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/go-chi/render"
+
+	"github.com/France-ioi/AlgoreaBackend/app/database"
+	"github.com/France-ioi/AlgoreaBackend/app/service"
+)
+
+// swagger:operation POST /contests/{item_id}/groups/{group_id} contests groups contestEnter
+// ---
+// summary: Enter the contest
+// description: >
+//                Allows to enter a contest as a group (user self or team).
+//
+//
+//                Restrictions:
+//                  * `item_id` should be a timed contest;
+//                  * `group_id` should be either the current user's self group (if the item's `has_attempts` is false) or
+//                     a team with `team_item_id` = `item_id` (otherwise);
+//                  * the authenticated user should have at least grayed access to the item;
+//                  * the authenticated user should be a member of the `group_id` (if it is a team);
+//                  * the group must be qualified for the contest (contestGetQualificationState returns "ready")
+//
+//                Otherwise, the "Forbidden" response is returned.
+// parameters:
+// - name: item_id
+//   description: "`id` of a contest"
+//   in: path
+//   type: integer
+//   format: int64
+//   required: true
+// - name: group_id
+//   in: path
+//   type: integer
+//   format: int64
+//   required: true
+// responses:
+//   "201":
+//     "$ref": "#/responses/contestEnterResponse"
+//   "401":
+//     "$ref": "#/responses/unauthorizedResponse"
+//   "403":
+//     "$ref": "#/responses/forbiddenResponse"
+//   "500":
+//     "$ref": "#/responses/internalErrorResponse"
+func (srv *Service) enter(w http.ResponseWriter, r *http.Request) service.APIError {
+	apiError := service.NoError
+	var qualificationState *contestGetQualificationStateResponse
+	var itemInfo struct {
+		Now                        *database.Time
+		Duration                   string
+		ContestParticipantsGroupID *int64
+	}
+	err := srv.Store.InTransaction(func(store *database.DataStore) error {
+		qualificationState, apiError = srv.getContestInfoAndQualificationStateFromRequest(r, store, true)
+		if apiError != service.NoError {
+			return apiError.Error
+		}
+
+		if qualificationState.State != ready {
+			apiError = service.InsufficientAccessRightsError
+			return apiError.Error
+		}
+
+		service.MustNotBeError(store.Items().ByID(qualificationState.itemID).
+			Select("NOW() AS now, items.duration, items.contest_participants_group_id").
+			WithWriteLock().Take(&itemInfo).Error())
+
+		service.MustNotBeError(store.Exec(`
+			INSERT INTO contest_participations (group_id, item_id, entered_at)
+			VALUES(?, ?, ?)`, qualificationState.groupID, qualificationState.itemID, itemInfo.Now).Error())
+
+		if itemInfo.ContestParticipantsGroupID != nil {
```

Comment:
I don't know what to think about the possibility of itemInfo.ContestParticipantsGroupID == nil.
I think we cannot allow that in our system, even if it does not destroy anything or give extra perm in practice. 
In order to be able to detect the problem, we should "log.warn" the case for the moment.